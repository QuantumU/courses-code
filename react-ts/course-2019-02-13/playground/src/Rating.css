.star-active {
  color: gold;
}

.star-inactive {
  color: lightgrey;
}