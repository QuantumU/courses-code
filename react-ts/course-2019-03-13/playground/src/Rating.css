.star {
  font-size: 3rem;
  color: lightgrey;
  user-select: none;
}

.star.active {
  color: gold;
}
