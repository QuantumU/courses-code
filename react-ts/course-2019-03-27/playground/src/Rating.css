.star-inactive {
  color: lightgrey;
  font-size: 40px;
}

.star-active {
  color: gold;
  font-size: 40px;
}