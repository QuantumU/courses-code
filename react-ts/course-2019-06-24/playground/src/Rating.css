.star {
  font-size: 30px;
  color: lightgrey;
  user-select: none;
}

.star.active {
  color: gold;
}
