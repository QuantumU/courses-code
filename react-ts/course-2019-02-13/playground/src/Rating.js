import React, { Component } from 'react';
import './Rating.css';

export default class Rating extends Component {
  /*
    props:
      stars: number
    events:
      onStarsChange: function(number)
  */
  render() {
    const starData = [];
    for (let i = 1; i <= 5; i++) {
      starData.push({
        id: i,
        active: i <= this.props.stars
      });
    }
    return (
      <div>
        {starData.map(star => (
          <span
            key={star.id}
            className={star.active ? 'star-active' : 'star-inactive'}
            onClick={() => {this.props.onStarsChange(star.id)}}
          >
            {star.active ? '★' : '☆'}
          </span>
        ))}
      </div>
    );
  }
}
