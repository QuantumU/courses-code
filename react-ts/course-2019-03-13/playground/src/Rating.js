import React, { Component } from 'react';

import './Rating.css';

class Rating extends Component {
  // props:
  //   stars: number
  // events:
  //   onStarsChange: (stars: number) => void

  shouldComponentUpdate(nextProps, nextState) {
    return this.props.stars !== nextProps.stars;
  }

  render() {
    const starIndexes = [1, 2, 3, 4, 5];

    return (
      <div>
        {starIndexes.map(index =>
          this.props.stars < index ? (
            <span
              className="star"
              key={index}
              onClick={() => {
                this.props.onStarsChange(index);
              }}
            >
              ☆
            </span>
          ) : (
            <span
              className="star active"
              key={index}
              onClick={() => {
                this.props.onStarsChange(index);
              }}
            >
              ★
            </span>
          )
        )}
      </div>
    );
  }
}

export default Rating;
