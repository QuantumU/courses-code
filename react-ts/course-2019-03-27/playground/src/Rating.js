import React from 'react';
import './Rating.css';
import {range} from 'lodash'

const Rating = React.memo(props => {
  /**
   * props:
   *   stars: number
   * event:
   *   onStarsChange: (newStars: number) => undefined
   */
  const starIndexes = range(1, 6);

  return starIndexes.map(index => (
    <span
      className={index <= props.stars ? 'star-active' : 'star-inactive'}
      onClick={() => {
        props.onStarsChange(index);
      }}
      key={index}
    >
      {index <= props.stars ? '★' : '☆'}
    </span>
  ));
});

export default Rating;
