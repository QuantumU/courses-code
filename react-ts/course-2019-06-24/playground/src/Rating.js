import React from "react";

import "./Rating.css";

const Rating = props => {
  const starIds = [1, 2, 3, 4, 5];

  if (![1, 2, 3, 4, 5].includes(props.stars)) {
    throw new Error("Invalid number of stars");
  }

  return (
    <div>
      {starIds.map(starId => {
        const isActive = starId <= props.stars;
        return (
          <span
            key={starId}
            className={"star" + (isActive ? " active": "")}
            onClick={() => {
              props.onStarsChange(starId);
            }}
          >
            {isActive ? "★" : "☆"}
          </span>
        );
      })}
    </div>
  );
};

export default Rating;
