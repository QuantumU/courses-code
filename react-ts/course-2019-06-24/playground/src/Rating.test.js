import React from "react";
import TestRenderer from "react-test-renderer";

import Rating from "./Rating";

describe("general rendering", () => {
  it("renders without crashing", () => {
    const instance = TestRenderer.create(<Rating stars={1} />).root;
  });

  it("renders 5 spans", () => {
    const instance = TestRenderer.create(<Rating stars={3} />).root;
    const spans = instance.findAllByType("span");
    expect(spans.length).toEqual(5);
    expect(spans).toHaveLength(5);
  });

  it("renders 3 active stars", () => {
    const instance = TestRenderer.create(<Rating stars={3} />).root;
    const activeStars = instance.findAllByProps({ className: "star active" });
    expect(activeStars).toHaveLength(3);
  });
});

describe("events", () => {
  it("triggers an onStarsChange event", () => {
    const mockFn = jest.fn();
    const instance = TestRenderer.create(
      <Rating stars={3} onStarsChange={mockFn} />
    ).root;
    const fourthStar = instance.findAllByType("span")[3];
    // simuliere einen Mausklick
    fourthStar.props.onClick();
    expect(mockFn).toBeCalledWith(4);
  });
});

describe("errors", () => {
  it("throws an error if the number of stars is negative", () => {
    const errorThrowingFunction = () => {
      TestRenderer.create(<Rating stars={-1} />);
    };
    expect(errorThrowingFunction).toThrow();
  });
});

describe("snapshot tests", () => {
  it("renders correctly", () => {
    const tree = TestRenderer.create(<Rating stars={2} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
