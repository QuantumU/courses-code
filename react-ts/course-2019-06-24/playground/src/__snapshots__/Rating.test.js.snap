// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests renders correctly 1`] = `
<div>
  <span
    className="star active"
    onClick={[Function]}
  >
    ★
  </span>
  <span
    className="star active"
    onClick={[Function]}
  >
    ★
  </span>
  <span
    className="star"
    onClick={[Function]}
  >
    ☆
  </span>
  <span
    className="star"
    onClick={[Function]}
  >
    ☆
  </span>
  <span
    className="star"
    onClick={[Function]}
  >
    ☆
  </span>
</div>
`;
